### PR TITLE
macro: add support for viewing attachment objects.

### PR DESCRIPTION
**Pull request by ypperlig**
Macros have attachments, but the attachments property on a macro only gives the attachment ids.
This PR implements a method to fetch all attachment objects for a given macro:
```python
>>> attachments = list(client.macros.attachments(macro.id))
>>> print(attachments)
[MacroAttachment(id=321)]
```

### CODE REVIEW

**Comment by ypperlig** [comment]
Also added support for download / upload of attachment to a macro (or unassociated).

File: zenpy/lib/endpoint.py [hunk 1]
```diff
@@ -470,6 +470,10 @@ class EndpointFactory(object):
     links = PrimaryEndpoint('services/jira/links')
     macros = MacroEndpoint('macros')
     macros.apply = SecondaryEndpoint('macros/%(id)s/apply.json')
+    macros.attachments = SecondaryEndpoint('macros/%(id)s/attachments.json')
+    macros.attachments_download = SecondaryEndpoint('macros/attachments/%(id)s/content')
```
Comment (facetoe) [inline zenpy/lib/endpoint.py]: This doesn't seem to be used anywhere

File: zenpy/lib/request.py [hunk 2]
```diff
@@ -219,6 +220,15 @@ def post(self, fp, token=None, target_name=None, content_type=None):
...
                 target_name = target_name or fp.name
 
+        if hasattr(fp, 'name'):
+            # Path objects (pathlib.Path)
+            if fp.name == '':
+                raise ZenpyException("upload requires a target file name")
+
+            target_name = target_name or fp.name
+
+            fp = open(fp, 'rb')
```
Comment (facetoe) [inline zenpy/lib/request.py]: Just realized we have a file leak (existing issue) as we don't close the file. Could you close this file object after the post?

**Comment by facetoe** [comment]
Looks good. Added a couple minor comments then good to merge.

**Comment by ypperlig** [comment]
Thanks for the quick update :) Latest commit removes the unused endpoint and closes the file pointer in UploadRequest.post.

File: zenpy/lib/request.py [hunk 3]
```diff
@@ -231,8 +241,30 @@ def post(self, fp, token=None, target_name=None, content_type=None):
...
+                url = self.api._build_url(self.api.endpoint.attachments_upload_unassociated(id=None))
+                
+            response = self.api._post(url, data=data, files=files, content_type=content_type)
+            
+        else:
+            url = self.api._build_url(self.api.endpoint.upload(filename=target_name, token=token))
+            response = self.api._post(url, data=fp, payload={}, content_type=content_type)
+
+        if not hasattr(fp, "close"):
+            raise ZenpyException("File pointer '{}' doesn't have a 'close' function.".format(fp))
```
Comment (facetoe) [inline zenpy/lib/request.py]:
```
if hasattr(fp, "close"):
   fp.close()
```

Is probably better

Comment (ypperlig) [reply zenpy/lib/request.py]: I agree, probably easier to read. Changed to if/else and if not available an exception is raised to ensure that file pointers always have a close method.

File: zenpy/lib/request.py [hunk 4]
```diff
@@ -231,8 +241,30 @@ def post(self, fp, token=None, target_name=None, content_type=None):
...
+            response = self.api._post(url, data=data, files=files, content_type=content_type)
+            
+        else:
+            url = self.api._build_url(self.api.endpoint.upload(filename=target_name, token=token))
+            response = self.api._post(url, data=fp, payload={}, content_type=content_type)
+
+        if hasattr(fp, "close"):
+            fp.close()
+        else:
+            raise ZenpyException("File pointer '{}' doesn't have a 'close' function.".format(fp))
```
Comment (facetoe) [inline zenpy/lib/request.py]: Last little thing! No need to throw the exception here if it doesn't have a close method.

Comment (ypperlig) [reply zenpy/lib/request.py]: Removed

**Comment by facetoe** [comment]
Thanks!

**Pull request merged**